jax/image/scale.py

```diff
@@ -0,0 +1,194 @@
+    out = np.where(x >= 1., ((-0.5* x + 2.5) * x - 4.) * x + 2., out)
+    return np.where(x >= 2., 0., out)
+  return 2., kernel
+
+
+def _compute_spans(input_size: int, output_size: int,
+                   scale: float, translate: float,
+                   kernel: Tuple[float, Callable[[np.ndarray], np.ndarray]],
+                   antialias: bool) -> Tuple[np.ndarray, np.ndarray]:
+  """
+  Computes the locations and weights of the spans of a 1D input image.
+  Returns:
+    A `(starts, weights)` tuple of ndarrays, where `starts` has shape
+    `[output_size]` and `weights` has shape `[output_size, span_size]`.
+  """
+  radius, kernel_fn = kernel
+  inv_scale = 1. / scale
+  kernel_scale = max(inv_scale, 1.) if antialias else 1.
+  span_size = min(2 * int(math.ceil(radius * kernel_scale)) + 1, input_size)
+
+  sample_f = (np.arange(output_size) + 0.5) * inv_scale * (1. - translate)
+  span_start = np.ceil(sample_f - radius * kernel_scale - 0.5)
+  span_start = np.clip(span_start, 0, input_size - span_size)
+  kernel_pos = (span_start - sample_f)[:, None] + np.arange(span_size) + 0.5
+  weight = kernel_fn(kernel_pos / kernel_scale)
+  total_weight_sum = np.sum(weight, axis=1, keepdims=True)
+  weights = np.where(
+      np.abs(total_weight_sum) > 1000. * np.finfo(np.float32).min,
+      weight / total_weight_sum, 0)
+  return span_start.astype(np.int32), weights
+
+
+def _scale_and_translate(x, output_shape, scale, translate, kernel,
+                         antialias):
+  input_shape = x.shape
+  assert len(input_shape) == len(output_shape)
+  assert len(input_shape) == len(scale)
+  assert len(input_shape) == len(translate)
+  spatial_dims = np.nonzero(
+      np.not_equal(input_shape, output_shape) |
+      np.not_equal(scale, 1) |
+      np.not_equal(translate, 0))[0]
+  if len(spatial_dims) == 0:
+    return x
+  output_spatial_shape = tuple(np.array(output_shape)[spatial_dims])
+  indices = []
+  dim_weights = []
+  slice_shape = list(input_shape)
+  in_names = string.ascii_lowercase[:len(output_shape) + len(spatial_dims)]
+  out_names = list(string.ascii_lowercase[:len(output_shape)])
+  einstr = in_names
+  for i, d in enumerate(spatial_dims):
+    m = input_shape[d]
+    n = output_shape[d]
+    starts, weights = _compute_spans(m, n, scale[d], translate[d],
+                                     kernel, antialias=antialias)
+    starts = lax.broadcast_in_dim(starts, output_spatial_shape + (1,), (i,))
+    slice_shape[d] = weights.shape[1]
+    indices.append(starts.astype(np.int32))
+    dim_weights.append(weights.astype(x.dtype))
+    einstr += ",{}{}".format(in_names[len(output_shape) + i], in_names[d])
+    out_names[d] = in_names[len(output_shape) + i]
+  index = lax.concatenate(indices, len(output_spatial_shape))
+  dnums = lax.GatherDimensionNumbers(
+      offset_dims=tuple(range(len(output_shape))),
+      collapsed_slice_dims=(),
+      start_index_map=tuple(spatial_dims))
+  out = lax.gather(x, index, dnums, slice_shape)
```

Comment:
might be clearer to use numpy style indexing with brackets [], at least for someone who doesn't know XLA quite as well :)

Reply:
Actually this is a case where I think I cannot do that. For example, for a bicubic resize (without antialiasing) this is gathering 4x4 patches in a single gather, but I don't see a way to gather windows with NumPy's indexing, which only really seems to support gathering single elements or entire axes.

Reply:
You can do this by adding an extra dimension to the array indexers, e.g.,
```
In [13]: image = np.zeros((50, 60))

In [14]: x, y = np.meshgrid(np.arange(40), np.arange(50), indexing='ij')

In [15]: image[x, y].shape
Out[15]: (40, 50)

In [16]: image[np.stack([x, x + 1], axis=-1), np.stack([y, y + 1], axis=-1)].shape
Out[16]: (40, 50, 2)
```

Reply:
This isn't quite the same thing once lowered to `lax.gather`. Your formulation does gathers of multiple indices, whereas the formulation I'm using here does a single gather of a continuous span. I don't know whether it makes a difference in practice, but it certainly seems like something the compiler could benefit from knowing. And it makes the indices array smaller!